cassiopeia/preprocess/UMI_utils.py

```diff
@@ -322,6 +346,64 @@ def form_clusters(
     return clusters
 
 
+def form_clusters_bayesian(
+    als: List[pysam.AlignedSegment],
+    q_threshold: Optional[int] = None,
+    proportion: float = 0.05,
+) -> List[pysam.AlignedSegment]:
+    """Forms clusters from a list of aligned segments (reads) probabilistically.
+
+    This function has the same purpose as :func:`form_clusters`, but instead
+    of using hardcoded quality thresholds, takes a probabilisitc approach to
+    forming read clusters by treating the quality scores as probabilities (in
+    fact, they are actually encoded probabilities of error). The two optional
+    arguments specify a dynamic threshold to assign a specific sequence to
+    a cluster. Specifically, for a given cluster sequence, a sequence is
+    assigned to this cluster (consensus) if its "match probability" to this
+    consensus is <= max(min(match probability), `q_threshold` * (`proportion` * length of longest sequence))
+    The "match probability" of a sequence to a possible consensus is is the sum
+    of the quality values where they do not match (equivalent to negative log
+    probability that all mismatches were sequencing errors).
```

Comment:
It seems that the match probability is defined as a sum of log probabilities but the thresholds are defined as probabilities (e.g., `q_threshold * proportion * length`). It sounds like you should be taking the log of this threshold probability that way the match probabilities can be compared to your threshold. Are you doing this in `ngs`?

Reply:
Yes, this is what is done internally

Reply:
You can think of the `q_threshold` and `proportion` as the following.
Assign a read to a candidate consensus if
```
(sum of the quality values where the read does not match the consensus) <= (q_threshold * proportion * length)
```
So, you are conceptually saying I want to allow at most (proportion * length) mismatches, each of base quality q_threshold for all reads assigned to a consensus.

Reply:
Good to know! Maybe we can make this a bit more clear in the documentation then?

Reply:
Yep, I removed the max,min part because that is just an implementation detail to deal with cases where only one sequence is provided. Also added description to what `q_threshold` is (PHRED quality score).